### PR TITLE
implement helm upgrade action

### LINKED ISSUE #2806: ✨ improve helm actions
### What are you trying to do?

I want to add a new helm action, `helm.#Upgrade` which is more useful than install in pipelines that run often. Additionally, this action should have more options to control the deployment, such as setting flags and values.

The usage could look something like this:

```c
release: helm.#Upgrade & {
  kubeconfig: client.commands.kubeconfig.stdout
  workspace:  client.filesystem.".".read.contents
  name:       "vault"
  repo:       "https://helm.releases.hashicorp.com"
  chart:      "vault"
  version:    "0.20.1"
  namespace:  "sandbox"
  install:    true
  atomic:     true
  flags: ["--skip-crds"]
  values: ["values.base.yaml", "values.staging.yaml"]
  set: #"""
    global.domain=example.com
    global.image=docker.io/bluebrown/echoserver
    """#
}
```

### Why is this important to you?

I think the current implementation in alpha is missing some functionality to make it really useful.  

Especially, being able to set values is important. 

The current implementation also doesn't allow running it more than once, which isn't suitable for CI. Helm will complain that the chart is already installed. Therefore, `upgrade --install` is the preferred way to use in a pipeline.

I have deployed helm charts in the past with terraform and the azure pipeline task, both have a signature as shown above.

### How are you currently working around this?

I hand rolled my own `helm.#Upgrade` action.

### PR DESCRIPTION
**Pull request by bluebrown**
> **Note** This is a prototype implementaton. Feedback for improvement would be appreciated.

Resolves #2806 

### Changes

- add  `#Upgrade` action in `universe.dagger.io/alpha/kubernetes/helm`

Signed-off-by: Nico Braun <rainbowstack@gmail.com>

### CODE REVIEW

File: pkg/universe.dagger.io/alpha/kubernetes/helm/test/test.cue
Comment (bluebrown) [inline pkg/universe.dagger.io/alpha/kubernetes/helm/test/test.cue:38]: The filesystem here should be `client.filesystem."./testdata".read.contents`

**Comment by gerhard** [comment]
Hi @bluebrown, I like the direction of this. I also use `helm upgrade --install` when installing new packages, this approach is definitely common.

Let me take a better look at what needs to improve before we can get this change merged.

**Comment by bluebrown** [comment]
I think we should decide what flags to support as first class struct fields. I think `--username`, `--password` and `--create-namespace` would be good in addition to the ones I already have. Especially `--password` could make sense to set it to `dagger.#Secret`.

**Comment by gerhard** [comment]
Thanks for sticking with me on this @bluebrown. There are a bunch of things which I needed to track down before I could update this PR. My commit comments have all the details.

The last commit in particular has a help request. It's OK if you are not able to help out. By the way, this includes `--username` & `--password` flags which I think are a great idea.

This is what I intend to do next:
- Pull https://github.com/dagger/dagger/pull/2807/commits/58c2371fc8b0ddfdbb8f5569bb3c963f6f9275c1 into a separate PR
- Re-enable K8s tests - have been disabled in #936 and remain disabled since (Sept. 2021)
- Figure out what else is missing before we can merge this

To be honest, I didn't expect this PR to become as big as it is. While we could merge it as soon as the existing checks go green, I would like to make this PR higher quality - especially running the right tests. I would like to spend at least another day doing so. My aim is to get this merged by the end of this week, July 22nd. Even if we do not get all the way with kind & Kubernetes tests, at least we will have made good progress in that direction. How does that sound @bluebrown?

File: pkg/universe.dagger.io/alpha/kubernetes/helm/test/test.cue
Comment (gerhard) [reply pkg/universe.dagger.io/alpha/kubernetes/helm/test/test.cue]: Great catch! Without this change, I was getting `cycle error` when running the integration tests. This made me dig & discover:
1. We are installing the wrong version of CUE locally, in our dev environments. [`dagger` is currently pinned to a `0.4.1` fork](https://github.com/dagger/dagger/blob/v0.2.24/go.mod#L135). This has all the (many!) details behind it: https://github.com/dagger/dagger/issues/2318
2. We are not running integration tests for the Kubernetes packages since #936 (they would have caught the existing `cycle error` that you have spotted)

While https://github.com/dagger/dagger/pull/2807/commits/316ce0e6b4128bdb2d5f754c34c2352b4e80119b solves this immediate issue, the right fix is a bit more involved (restoring #936) and I intend to do it next, as a separate PR.

Once again, thanks for noticing this!

**Comment by bluebrown** [comment]
Hi @gerhard, as far as I understand, you want me to adjust the tests, to do a real deployment instead of dry-run.  Yes, I can do that.

I also would like thi add the `--atomic`, flag back to first class. I think its one of the most useful flags, at we use it always to ensure our apps keep running stable, in case something goes wrong during release.

**Comment by gerhard** [comment]
> Hi @gerhard, as far as I understand, you want me to adjust the tests, to do a real deployment instead of dry-run. Yes, I can do that.

Yes please. Single test that covers the happy path & ideally all the flags. If `--username` & `--password` are too much hassle, it's OK to skip those.

> I also would like to add the `--atomic`, flag back to first class. I think it's one of the most useful flags. At least, we use it always to ensure our apps keep running stable, in case something goes wrong during release.

Good catch! I removed too much.

There are also a bunch of package coding style elements that we are missing. [The public definitions](https://docs.dagger.io/1226/coding-style#all-public-definitions-and-fields-must-be-documented) will become highly useful with the imminent release of https://github.com/dagger/daggerlsp

File: pkg/universe.dagger.io/alpha/kubernetes/helm/upgrade.cue
Comment (jpadams) [inline pkg/universe.dagger.io/alpha/kubernetes/helm/upgrade.cue:14]: I'd like to see whatever we do around images be consistent across `upgrade.cue` and `build.cue`.

Could you also propose the same `image:` in build.cue? 🙏

Comment (bluebrown) [reply pkg/universe.dagger.io/alpha/kubernetes/helm/upgrade.cue:14]: I have added the image field to the `install.cue` action, assuming this is what you meant by `build.cue`.

**Review by gerhard — APPROVED** [review]
This is looking good - thank you for the great contribution @bluebrown!

As soon as the checks pass and the tests that I'm running locally pass, this will get merged.

**Comment by gerhard** [comment]
There is an issue with the image, still looking into this:

```
❯ dagger do test install URL

Dagger Cloud URL: https://dagger.cloud/runs/bc253954-0ccb-4fb3-8107-46b8c6c6738a

[✔] client.env                                                                                                       0.0s
[✔] client.commands.kubeconfig                                                                                       0.0s
[✗] actions.test.install.URL.run                                                                                     0.0s
5:19PM FATAL failed to execute plan: task failed: actions.test.install.URL.run._exec: invalid FS at path "actions.test.install.URL.run._exec.input": FS is not set
```

**Comment by gerhard** [comment]
image

image

⛴

**Pull request merged**
